include/sys/onoff.h

```diff
@@ -0,0 +1,611 @@
+/*
+ * Copyright (c) 2019 Peter Bigot Consulting, LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_ONOFF_H_
+#define ZEPHYR_INCLUDE_SYS_ONOFF_H_
+
+#include <kernel.h>
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup resource_mgmt_apis Resource Management APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief Flag fields used to specify on-off service behavior.
+ */
+enum onoff_service_flags {
+	/**
+	 * @brief Flag passed to onoff_service_init().
+	 *
+	 * When provided this indicates the start transition function
+	 * may cause the calling thread to wait.  This blocks attempts
+	 * to initiate a transition from a non-thread context.
+	 */
+	ONOFF_SERVICE_START_SLEEPS      = BIT(0),
+
+	/**
+	 * @brief Flag passed to onoff_service_init().
+	 *
+	 * As with @ref ONOFF_SERVICE_START_SLEEPS but describing the
+	 * stop transition function.
+	 */
+	ONOFF_SERVICE_STOP_SLEEPS       = BIT(1),
+
+	/**
+	 * @brief Flag passed to onoff_service_init().
+	 *
+	 * As with @ref ONOFF_SERVICE_START_SLEEPS but describing the
+	 * reset transition function.
+	 */
+	ONOFF_SERVICE_RESET_SLEEPS      = BIT(2),
+
+	/* Internal use. */
+	ONOFF_SERVICE_HAS_ERROR         = BIT(3),
+
+	/* This and higher bits reserved for internal use. */
+	ONOFF_SERVICE_INTERNAL_BASE     = BIT(4),
+};
+
+/* Forward declaration */
+struct onoff_service;
+
+/**
+ * @brief Signature used to notify an on-off service that a transition
+ * has completed.
+ *
+ * Functions of this type are passed to service-specific transition
+ * functions to be used to report the completion of the operation.
+ * The functions may be invoked from any context.
+ *
+ * @param srv the service for which transition was requested.
+ *
+ * @param res the result of the transition.  This shall be
+ * non-negative on success, or a negative error code.  If an error is
+ * indicated the service shall enter an error state.
+ */
+typedef void (*onoff_service_notify_fn)(struct onoff_service *srv,
+					int res);
+
+/**
+ * @brief Signature used by service implementations to effect a
+ * transition.
+ *
+ * Service definitions use two function pointers of this type to be
+ * notified that a transition is required, and a third optional one to
+ * reset service state.
+ *
+ * The start function will be called only from the off state.
+ *
+ * The stop function will be called only from the on state.
+ *
+ * The reset function may be called only when
+ * `onoff_service_has_error()` returns true.
+ *
+ * Functions will be invoked from interrupt context only if the
```

Comment:
No, that's a leftover because #21584 needs to get merged so `k_is_pre_kernel()` is documented and visible without having to include the system init header.  I'll revisit this when that's merged.

Reply:
>  I'll revisit this when that's merged.

That's been merged, so this is unblocked.